Bluetooth: AVRCP: connection and disconnection of AVRCP

gzh-terry opened the pull request:
This pull request implemented the basic AVRCP functions, which includes:
1. Establishing AVCTP channel.
2. Create AVRCP instance.
3. Register SDP services.
4. A shell function to test these interfaces.

This is the first part of the entire AVRCP functions, we tested the code by
establishing AVCTP channel with Android phones.

Four configs are provided to turn on/off the functions:
1. CONFIG_BT_AVRCP_TARGET
2. CONFIG_BT_AVRCP_CONTROLLER
3. CONFIG_BT_AVRCP
4. CONFIT_BT_AVCTP

gzh-terry commented:
Hi @lylezhu2012 ,

May I doublecheck if this is the correct method to start a new feature?
AVRCP would include a series of pull requests. Will it be better if I submit them together?

Best regards,
Zihao

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,151 @@
+	session = AVCTP_CHAN(chan);
+	LOG_DBG("chan %p session %p", chan, session);
+
+	session->ops->connected(session);
```
I think `session->ops` and `session->ops->connected` need to check before using it.

gzh-terry replied:
Yes, you are correct. Would fix it.

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,151 @@
+	LOG_DBG("chan %p session %p", chan, session);
+	session->br_chan.chan.conn = NULL;
+
+	session->ops->disconnected(session);
```
Same here, `session->ops` and `session->ops->disconnected` need to check before using it.

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,151 @@
+{
+	/* TODO */
+
+	return 0;
```
Please return an error code, such as `-ENOTSUP`, instead of `0` if the feature is unsupported.

gzh-terry replied:
Would change it to -ENOTSUP.

The process of AVCTP packets are included from the next pull request.
BTW I think it is still fine if an L2CAP PDU is received correctly but nothing happens.

lylezhu2012 replied:
I don't see any changes here. Please confirm.

gzh-terry replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,151 @@
+
+#define AVCTP_CHAN(_ch) CONTAINER_OF(_ch, struct bt_avctp, br_chan.chan)
+
+K_MUTEX_DEFINE(avctp_mutex);
+#define AVCTP_LOCK() k_mutex_lock(&avctp_mutex, K_FOREVER)
+#define AVCTP_UNLOCK() k_mutex_unlock(&avctp_mutex)
```
I did not find the lock is used anywhere. And for `avctp_mutex`, if it is internal only, I recommend defining it as a static variable.

gzh-terry replied:
Removed temporarily, and will add it back when actually needed.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,369 @@
+
+#define AVRCP_AVCTP(_avctp) CONTAINER_OF(_avctp, struct bt_avrcp, session)
+
+K_MUTEX_DEFINE(avrcp_mutex);
```
Same here. For `avrcp_mutex`, if it is internal only, I recommend defining it as a static variable.

gzh-terry replied:
Add static here.

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,366 @@
+		return NULL;
+	}
+
+	for (i = 0; i < CONFIG_BT_MAX_CONN; i++) {
```
Please use `ARRAY_SIZE(xxxx)` instead of `CONFIG_BT_MAX_CONN`.

gzh-terry replied:
Done

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,366 @@
+		return NULL;
+	}
+
+	for (i = 0; i < CONFIG_BT_MAX_CONN; i++) {
```
Same here.

gzh-terry replied:
Done

jhedberg commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,149 @@
+	session = AVCTP_CHAN(chan);
+	LOG_DBG("chan %p session %p", chan, session);
+
+	if (session->ops && session->ops->connected)
+		session->ops->connected(session);
```
Missing `{ }`

gzh-terry replied:
Done

jhedberg commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,149 @@
+	LOG_DBG("chan %p session %p", chan, session);
+	session->br_chan.chan.conn = NULL;
+
+	if (session->ops && session->ops->disconnected)
+		session->ops->disconnected(session);
```
Missing `{ }`

gzh-terry replied:
Done

jhedberg commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,149 @@
+	session->br_chan.chan.ops = &ops;
+	session->br_chan.required_sec_level = BT_SECURITY_L2;
+
+	return bt_l2cap_chan_connect(conn, &session->br_chan.chan,
+		BT_L2CAP_PSM_AVCTP);
```
Incorrect alignment, however I think this might just fit onto one line

gzh-terry replied:
Done

jhedberg commented on include/zephyr/bluetooth/classic/avctp.h:
I'd just remove this header file. I suspect we'll never need a public interface for AVCTP. If we end up having such a need then we can always add it later.

gzh-terry replied:
This make sense.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,365 @@
+static struct bt_sdp_record avrcp_ct_rec = BT_SDP_RECORD(avrcp_ct_attrs);
+#endif /* CONFIG_BT_AVRCP_CONTROLLER */
+
+static struct bt_avrcp *get_new_connection(struct bt_conn *conn)
+{
+	uint8_t i;
+
+	if (!conn) {
+		LOG_ERR("Invalid Input (err: %d)", -EINVAL);
+		return NULL;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(avrcp_connection); i++) {
+		if (!avrcp_connection[i].session.br_chan.chan.conn) {
+			memset(&avrcp_connection[i], 0, sizeof(avrcp_connection[i]));
+			return &avrcp_connection[i];
+		}
+	}
+
+	LOG_ERR("More connection cannot be supported");
+	return NULL;
+}
+
+static struct bt_avrcp *avrcp_lookup_by_conn(struct bt_conn *conn)
+{
+	uint8_t i;
+
+	if (!conn) {
+		return NULL;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(avrcp_connection); i++) {
+		if (avrcp_connection[i].session.br_chan.chan.conn == conn) {
+			return &avrcp_connection[i];
+		}
+	}
+
+	return NULL;
+}
```
You should be able to make all of this much simpler by using the `bt_conn_index()` API. That returns an array index that can be used whenever you have an array with `CONFIG_BT_MAX_CONN` elements. You get the AVRCP context then with:
```c
&avrcp_connection[bt_conn_index(conn)]
```

gzh-terry replied:
> You should be able to make all of this much simpler by using the `bt_conn_index()` API. That returns an array index that can be used whenever you have an array with `CONFIG_BT_MAX_CONN` elements. You get the AVRCP context then with:
> 
> ```c
> &avrcp_connection[bt_conn_index(conn)]
> ```

Thanks, this really makes the code clean.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,365 @@
+			},
+			{
+				BT_SDP_TYPE_SIZE(BT_SDP_UINT16),
+				BT_SDP_ARRAY_16(0X0104u)
```
What's the `0x0104` magic number? Maybe add a define for it?

gzh-terry replied:
Yes, this is the AVCTP version, V14 is the latest version now.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,365 @@
+			},
+			{
+				BT_SDP_TYPE_SIZE(BT_SDP_UINT16),
+				BT_SDP_ARRAY_16(0x0106U)
```
Same, 1.6.2 is the latest AVRCP version.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,365 @@
+		},
+		)
+	),
+	BT_SDP_SUPPORTED_FEATURES(0x0003U), /* Category 1 & 2 */
```
Done

gzh-terry commented:
Hi @lylezhu2012 ,
Would you mind taking a look at the latest version?

Best regards,
Zihao

jhedberg commented:
@gzh-terry you need to rebase to fix the merge conflict

jhedberg commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -0,0 +1,88 @@
+ *
+ *  @return 0 in case of success or error code in case of error.
+ */
+int bt_avrcp_register_cb(struct bt_avrcp_cb *cb);
```
Since this is a singleton (e.g. you don't have a linked list pointer in the struct), I think `const struct bt_avrcp_cb *cb` is probably more correct. That said, have you considered whether you want to keep this a singleton, i.e. that there aren't any valid use cases where an app might be interested in registering multiple instances of the callbacks?

gzh-terry replied:
Hmm, technically the Bluetooth Spec allows multiple instances, however, none of the applications did so.
As for the target side, one AVRCP instance can control multiple players, thus multiple instances is meaningless.
As for the controller side, we don't really need two distinct control panels on one product.
I saw both A2DP and HFP assume singleton, I choose to follow. 'const' is not used as well at these profiles.

jhedberg replied:
> I saw both A2DP and HFP assume singleton, I choose to follow. 'const' is not used as well at these profiles.

We should then add `const` to those places too. Please don't assume that existing code must be "correct" just because it's already in the tree, especially in the case of Bluetooth Classic.

gzh-terry replied:
> > I saw both A2DP and HFP assume singleton, I choose to follow. 'const' is not used as well at these profiles.
> 
> We should then add `const` to those places too. Please don't assume that existing code must be "correct" just because it's already in the tree, especially in the case of Bluetooth Classic.

Done.

jhedberg commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -178,6 +178,30 @@ config BT_A2DP_SINK
+	help
+	  This option enables the AVRCP profile
+
+config BT_AVRCP_TARGET
+	bool "Bluetooth AVRCP Profile Target Function"
+	depends on BT_AVRCP
+	help
+	  This option enables the AVRCP profile target function
+
+config BT_AVRCP_CONTROLLER
+	bool "Bluetooth AVRCP Profile Controller Function"
+	depends on BT_AVRCP
+	help
+	  This option enables the AVRCP profile controller function
```
I'd suggest removing the `depends on BT_AVRCP` and put these instead inside a `if BT_AVRCP` ... `endif` block. That way you can add more `BT_AVRCP_*` options in the future without having to declare `depends on BT_AVRCP` every time.

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -0,0 +1,70 @@
+
+typedef enum __packed {
+	BT_AVCTP_PKT_TYPE_SINGLE = 0b00,
+} bt_avctp_pkt_typ_t;
```
s/typ/type/ (the enum value uses "TYPE" so you should be consistent about this)

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -0,0 +1,70 @@
+struct bt_avctp_header {
+	uint8_t ipid: 1;    /* Invalid Profile Identifier (1), otherwise zero (0) */
+	uint8_t cr: 1;      /* Command(0) or Respone(1) */
+	uint8_t pkt_typ: 2; /* Set to zero (00) for single L2CAP packet */
```
same here

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -0,0 +1,70 @@
+int bt_avctp_disconnect(struct bt_avctp *session);
+
+/* Create AVCTP PDU */
+struct net_buf *bt_avctp_create_pdu(struct bt_avctp *session,
+	bt_avctp_cr_t cr, bt_avctp_ipid_t ipid, uint8_t *tid, uint16_t pid);
```
The split line isn't correctly aligned.

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,467 @@
+#define AVRCP_TIMEOUT K_SECONDS(3) /* Shell be greater than TMTP (1000ms) */
+#define AVRCP_AVCTP(_avctp) CONTAINER_OF(_avctp, struct bt_avrcp, session)
+#define AVRCP_KWORK(_work) CONTAINER_OF(CONTAINER_OF(_work, struct k_work_delayable, work),\
+										struct bt_avrcp, timeout_work)
```
too long line and also not correctly aligned

gzh-terry replied:
Done.

jhedberg commented on subsys/bluetooth/host/classic/l2cap_br.c:
```diff
@@ -2074,9 +2076,17 @@ void bt_l2cap_br_init(void)
 		bt_avdtp_init();
 	}
 
+	if (IS_ENABLED(CONFIG_BT_AVCTP)) {
+		bt_avctp_init();
+	}
+
 	bt_sdp_init();
 
 	if (IS_ENABLED(CONFIG_BT_A2DP)) {
 		bt_a2dp_init();
 	}
+
+	if (IS_ENABLED(CONFIG_BT_AVRCP)) {
+		bt_avrcp_init();
+	}
```
This isn't particularly scalable. It should be possible to implement new profiles/protocols completely on the application side, which means that you can't have a requirement that the stack-internal `l2cap_br.c` needs to be modified every time you add some new profile. The profiles/protocols which are part of the Bluetooth core specification (L2CAP, SDP, etc) could be treated as an exception, but for everything else you need to find a solution which doesn't require modifying this internal c-file.

jhedberg commented on subsys/bluetooth/host/classic/shell/avrcp.c:
```diff
@@ -0,0 +1,155 @@
+#include "host/shell/bt.h"
+
+struct bt_avrcp *default_avrcp;
+static uint8_t avrcp_registered;
```
Why is this not a `bool`?

gzh-terry replied:
Copied the style from a2dp.c :)
Anyway, it can be a bool.

gzh-terry replied:
Done.

gzh-terry commented on subsys/bluetooth/host/classic/avctp_internal.h:
```diff
@@ -0,0 +1,70 @@
+	uint8_t ipid: 1;    /* Invalid Profile Identifier (1), otherwise zero (0) */
+	uint8_t cr: 1;      /* Command(0) or Respone(1) */
+	uint8_t pkt_typ: 2; /* Set to zero (00) for single L2CAP packet */
+	uint8_t tid: 4;     /* Transaction label */
```
I'm thinking about this, but it's really complecated when we have read and write macro for all these fields..

gzh-terry replied:
I've seen many other files have a bit field definition in a similar way, for example, just search code " : 1;".
Does this really matter?

jhedberg replied:
If those places are used for mapping to well defined protocol payloads, then the code is likely buggy and may fail with certain architectures or compilers.

Take a look at e.g. https://stackoverflow.com/questions/6043483/why-bit-endianness-is-an-issue-in-bitfields or if you want to go to the authoritative source (the C standard), take a look at section 6.7.2.1/11 here: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2310.pdf. I can quote the relevant text:

> The order of allocation of bit-fields within a unit (high-order to low-order or low-order to high-order) is implementation-defined.

In practice what I've seen is that all big-endian architectures do it one way and little-endian the other, and I've seen code which have build-time conditional variants of struct definitions depending on the host endianness, however considering the above from the C standard, even such solutions are prone to be buggy.

jhedberg replied:
> I'm thinking about this, but it's really complecated when we have read and write macro for all these fields..

Zephyr provides some helpers for this in `include/zephyr/sys/util_macro.h`. Take a look at e.g. `FIELD_PREP()` and `FIELD_GET()` as well as how they are used throughout the tree.

jhedberg replied:
Similar discussion going on here: https://github.com/zephyrproject-rtos/zephyr/pull/79497#discussion_r1799557709

gzh-terry replied:
Very solid reasons.

I've changed it.

gzh-terry commented:
Hi @lylezhu2012 ,

I've add some Tx/Rx code since last update, would you mind taking a look one more time?

Best regards,

lylezhu2012 commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,460 @@
+	return 0;
+}
+
+int bt_avrcp_get_unit_info(struct bt_avrcp *avrcp)
+{
+	struct net_buf *buf;
+	uint8_t param[5];
+
+	buf = avrcp_create_unit_pdu(avrcp, BT_AVCTP_CMD);
+	if (!buf) {
+		return -ENOMEM;
+	}
+
+	memset(param, 0xFF, 5);
+	net_buf_add_mem(buf, param, sizeof(param));
+
+	return avrcp_send(avrcp, buf);
+}
```
Is there something missing in the functionality of this function? I don't see anywhere handling the response of this command. From Api perspective, the result should be notified if the function is called from application.

gzh-terry replied:
Not really now, since the unit info is actually fixed in AVRCP.
As said in AVRCP v1.6.2, the TG device shall return a unit_type PANEL(0x09) in the response frame.
In the mojority cases, we do not care about this return value, or we can add a error handling if the response is incorrect.

gzh-terry replied:
Done.

gzh-terry commented on subsys/bluetooth/host/classic/Kconfig:
```diff
@@ -178,6 +178,32 @@ config BT_A2DP_SINK
+config BT_AVCTP
+	bool "Bluetooth AVCTP protocol support"
+	help
+	  This option enables Bluetooth AVCTP support
```
Done.

gzh-terry commented on include/zephyr/bluetooth/classic/avrcp.h:
```diff
@@ -0,0 +1,88 @@
+ */
+int bt_avrcp_register_cb(const struct bt_avrcp_cb *cb);
+
+/** @brief Disconnect AVRCP Unit Info.
```
Done.

gzh-terry commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,221 @@
+static void avctp_l2cap_disconnected(struct bt_l2cap_chan *chan)
+{
+	struct bt_avctp *session = AVCTP_CHAN(chan);
+
```
Done.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,460 @@
+		return -ENOMEM;
+	}
+
+	memset(param, 0xFF, 5);
```
Done.

gzh-terry commented on subsys/bluetooth/host/classic/avrcp.c:
```diff
@@ -0,0 +1,460 @@
+	bt_avrcp_subunit_id_t subunit_id;
+	bt_avrcp_subunit_type_t subunit_type;
+
+	avctp_hdr = (void *)buf->data;
+	net_buf_pull(buf, sizeof(*avctp_hdr));
+	avrcp_hdr = (void *)buf->data;
```
Done.

gzh-terry commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,221 @@
+{
+	struct net_buf *rsp;
+	struct bt_avctp *session = AVCTP_CHAN(chan);
+	struct bt_avctp_header *hdr = (void *)buf->data;
```
Fragmented AVCTP packets are not supported now.
Would do the re-assemble later.

gzh-terry replied:
I've added error handler for the fragmented ones.

gzh-terry commented on subsys/bluetooth/host/classic/avctp.c:
```diff
@@ -0,0 +1,221 @@
+{
+	struct net_buf *rsp;
+	struct bt_avctp *session = AVCTP_CHAN(chan);
+	struct bt_avctp_header *hdr = (void *)buf->data;
```
Done.

gzh-terry committed:
Bluetooth: AVCTP: Implementation of AVCTP.

This patch implementing avctp.c
New Kconfig BT_AVCTP is provided to enable this layer.
avctp_internal.h shows the APIs for the upper layer, i.e., AVRCP.

Only connection and disconnection interfaces are provided in this patch.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVRCP: Add SDP attributes.

This patch add SDP records for both CT and TG role.

The SDP attribute would be registered according to the configuration.

OBEX and Browsing commands are optional and yet not supported.

SDP registration is implemented at AVRCP level to simplify the
workload of the upper layer. We assume the App can have limited
knowledge on SDP structures.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: Shell: add command to obtain unit info

This patch allow to acquire the unit info of the remote device.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVRCP: Implemation of AVRCP.

This patch implementing avrcp.c
New Kconfig BT_AVRCP is provided to enable this layer.
BT_AVRCP_TARGET and BT_AVRCP_CONTROLLER are then
provided to enable one of the two roles independently.
avrcp.h shows the APIs for the upper layer.

Only connection and disconnection interfaces are provided in this patch.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVRCP: add shell tools for AVRCP functions.

Only the basic functions for establishing an AVCTP connection
are provided at this stage.
An BR/EDR ACL connection is necessary before AVRCP function.
Register callbacks before utilizing AVRCP.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth : AVRCP: allow to receive an AVRCP message.

This patch received an AVRCP message and remove timeout timers.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVCTP: allow to create and send AVCTP message

This patch defines the message format for general AVCTP.
They would be further called by AVRCP layer.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVRCP: allow to create and send AVRCP unit message

This patch defines the message format for AVCTP unit message.
This is the first out of the four types of commands and can be
used by the CT to obtain the unit info from the TG device.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVCTP: allow to receive an AVCTP message.

This patch received an AVCTP message and forward to the
upper layer, e.g., AVRCP.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVRCP: fix bitfield issue.

The bit order can be incorrect when use bit field definition.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVCTP: check buffer length before use.

The buffer length shall be validated before extracting
Transaction ID and C/R field.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVCTP: add error handling for fragmented message.

Fragmented AVCTP message is not supported now. Therefore
A error message is printed.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Bluetooth: AVRCP: allow to parse unit info resposne.

Add AVRCP handler for UNIT INFO responses.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry commented:
Hi @lylezhu2012 

Changes have been made since the last review. Would you mind taking a look to see if it's fine with you?

Best regards,
